File: src/feature/benchmark/mfcc-benchmark.lua
--[[
Copyright (c) Facebook, Inc. and its affiliates.
All rights reserved.

This source code is licensed under the BSD-style license found in the
LICENSE file in the root directory of this source tree.
--]]

local speech = require 'speech'

local samplerate = 16000
local fs = samplerate
local tw = 25
local ts = 10
local M = 20
local L = 22
local dev = 2
local mel_floor = 1
local R1 = 0
local R2 = samplerate/2
local alpha = 0.97
local N = 13
local T = 8 -- seconds

local preprocess = speech.Mfcc{
   fs=fs, tw=tw, ts=ts, M=M, N=N, L=L, dev=dev, mel_floor=mel_floor,
   alpha=alpha, R1=R1, R2=R2
}
local o
local x = torch.rand(T * samplerate)
for i=1,10 do
  o = preprocess(x) -- warmup
end
local timer = torch.Timer()
for i=1,100 do
  o = preprocess(x)
end
local t = timer:time().real
print(o:sum())
print(string.format("Average time taken  %.5f milliseconds\n",
    t * 1000.0 / 100));

File: src/feature/benchmark/MfccBenchmark.cpp
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <algorithm>
#include <chrono>
#include <iomanip>
#include <iostream>
#include <utility>

#include "libraries/feature/Mfcc.h"

using namespace w2l;

int main() {
  FeatureParams params;
  params.samplingFreq = 16000;
  params.frameSizeMs = 25;
  params.frameStrideMs = 10;
  params.numFilterbankChans = 20;
  params.lowFreqFilterbank = 0;
  params.highFreqFilterbank = 8000;
  params.numCepstralCoeffs = 13;
  params.lifterParam = 22;
  params.deltaWindow = 2;
  params.accWindow = 2;
  params.zeroMeanFrame = false;
  params.useEnergy = false;
  params.usePower = false;
  params.windowType = WindowType::HANNING;
  Mfcc<float> mfcc(params);

  std::vector<int> audiotimesec = {1, 10, 15, 20, 25, 50, 100}; // in seconds
  int64_t ntimes = 1000;

  std::cout << "Benchmark MFCC" << std::endl;
  for (int64_t t : audiotimesec) {
    double total_time = 0.0;
    std::vector<float> input(t * params.samplingFreq);
    for (int64_t M = 0; M < ntimes; M++) {
      std::generate(input.begin(), input.end(), []() {
        return (rand() * 1.0 / RAND_MAX);
      });
      auto start = std::chrono::system_clock::now();
      auto output = mfcc.apply(input);
      auto end = std::chrono::system_clock::now();
      total_time +=
          (std::chrono::duration_cast<std::chrono::milliseconds>(end - start))
              .count();
    }
    std::cout << "| Input Size : " << t << " sec , Avg. time taken "
              << std::setprecision(5) << total_time / ntimes << " msec"
              << std::endl;
  }
  return 0;
}
